src/checkpoint/checkpoint_api.h

```diff
@@ -51,29 +51,93 @@
 
 namespace checkpoint {
 
+/// Callback for user to allocate bytes during serialization
 using BufferCallbackType = std::function<char*(std::size_t size)>;
 
+/// Return of serialize that contains the buffer and size serialized
 struct SerializedInfo {
   virtual std::size_t getSize() const = 0;
   virtual char* getBuffer() const = 0;
   virtual ~SerializedInfo() { }
 };
 
-using SerializedInfoPtrType = std::unique_ptr<SerializedInfo>;
-using SerializedReturnType = SerializedInfoPtrType;
+/// Convenience typedef for \c std::unique_ptr<SerializedInfo>
+using SerializedReturnType = std::unique_ptr<SerializedInfo>;
 
+/**
+ * \brief Serialize \c T into a byte buffer
+ *
+ * Serializes an object recursively by invoking the \c serialize
+ * functions/methods recursively.
+ *
+ * \param[in] target the \c T to serialie
+ * \param[in] fn (optional) callback to supply buffer for to allow user
+ * allocation of the produced byte buffer. The callback will pass the number of
+ * bytes required and return a char* will at least that many bytes.
+ *
+ * \return a \c std::unique_ptr to a \c SerializedInfo containing the buffer
+ * with serialized data and the size of the buffer
+ */
 template <typename T>
 SerializedReturnType serialize(T& target, BufferCallbackType fn = nullptr);
 
+/**
+ * \brief De-serialize and reify \c T from a byte buffer and corresponding \c
+ * size
+ *
+ * De-serializes an object recursively by first invoking the reconstruction
+ * strategy and then \c serialize functions/methods recursively to rebuild the
+ * state of the object as serialized. During reconstruction, based on trait
+ * detection, \c T will either be default constructed on \c user_buf (or a
+ * system allocated buffer) or reconstructed based on a user-defined reconstruct
+ * method. If \c user_buf is not passed, the returned object point must be
+ * deallocated with \c delete
+ *
+ * \param[in] buf the buffer containing the bytes to reify \c T
+ * \param[in] len the number of bytes in \c buf
+ * \param[in] user_buf (optional) buffer containing bytes allocated with
+ * sufficient size for \c T. If this buffer is passed, the caller is responsible
+ * for deallocating the buffer. If it is not passed, the system will allocate a
+ * buffer that must be de-allocated with \c delete
+ *
+ * \return a pointer to the newly reified \c T based on bytes in \c buf
+ */
 template <typename T>
-T* deserialize(char* buf, std::size_t size, T* user_buf = nullptr);
+T* deserialize(char* buf, std::size_t len, T* user_buf = nullptr);
```

Comment:
Would it make sense to have two distinct overloads here
```
T* deserialize(char* data_buf, std::size_t len, char* object_buf);
std::unique_ptr<T> deserialize(char* buf, std::size_t len);
```
to distinguish the ownership cases?

I think the second case is fully sound.

In the first case, I'm conflicted as to what type `object_buf` should actually have, and whether the function should return `T*` or `void`. At the point of the call, `object_buf` is not a "pointer to `T`", which would argue for accepting `char*` or `void*`. Returning `T*` would be a convenience to avoid ceremony in a call like

`T* t = deserialize(buf, len, malloc(sizeof(T)));`

But would feel weird in the case of 

`deserialize(buf, len, &array_of_t[i]);`

Reply:
Ok, added a unique_ptr interface